### ISSUE TITLE
Evolve sequentially (a snaking path)

### ISSUE THREAD

**Issue by schrum2** [schrum2/EvoCraft-SCOPE#23]
Make a command line parameter EVOLVE_SNAKE that completely changes the way the CPPNs generate structures if set to True.

In this mode of generation, the CPPN has an additional 6 outputs (one for each direction you can move), and then one final output after that for a total of 7. The final continuation output is for whether or not to continue generating the shape.

First query the CPPN at scaled coordinates (0,0,0) and add whatever resulting Block to a list of Blocks that will eventually be generated. Then check the continuation output, and if it is high enough, keep going (otherwise stop generating blocks). If continuing, check the 6 directional outputs and figure out which one is the largest (each corresponds to adding or subtracting one from one of the three coordinates). Query the CPPN at the new coordinates and repeat the process until the continuation output is low enough to stop.

Note that there are three sets of coordinates here: World, Relative Placement, and Query. The World coordinates are the actual global coordinates where things will be placed. We want to place multiple shapes in different places, so I was not talking about the World coordinates above. The Relative Placement coordinates are the same scale of World coordinates (one unit = one block size), but are defined relative to wherever the center of the generated structure is supposed to be. So, if the shape were generated between global coordinates x = 10 and x = 16 then the 0 Relative Placement in the x-dimension would be 13 globally. The last coordinate system is for CPPN Queries. If you wanted to query x = 1 in the example above, then the CPPN input for x would be 0.333333 because 13 + 1 = 14 is 1/3 of the way to 16.

Some other comments:
- It is ok for a shape to snake outside the bounds of where it should be generated. Let's see what happens and make decisions from there.
- We need a way to avoid infinite loops. Make a command line parameter called MAX_SNAKE_LENGTH with a default of 100. If the continuation output has not signaled a stop by this point, then the loop will exit anyway.

**Comment by schrum2** [comment]
I think I've confirmed the infinite loop. Now to troubleshoot. However, I also added  a way to clear out the parts of the snakes that go beyond the bounds of what is normally cleared.

**Comment by schrum2** [comment]
Be sure to clean up all the debugging print statements eventually @richeyme

**Comment by schrum2** [comment]
I'm not sure that the error we are experiencing is an infinite loop. It seems to correspond to an error happening on the server end. The server throws a Java exception that looks like this
```
[09:43:44 ERROR] [Sponge]: The Scheduler tried to run the task spawnBlocks owned by Plugin{id=minecraft_rpc, name=Minecraft RPC, version=1.0, description=A plugin for Sponge which lets you control Minecraft using gRPC., source=mods\minecraft-rpc-0.0.5.jar}, but an error occured.
org.spongepowered.api.util.PositionOutOfBoundsException: Position is out of bounds: expected in range (-30000000, 0, -30000000) to (29999999, 255, 29999999) but got (92, -1, 5)
        at net.minecraft.world.World.checkBlockBounds(SourceFile:3568) ~[amu.class:?]
        at net.minecraft.world.WorldServer.setBlock(SourceFile:1525) ~[oo.class:?]
        at net.minecraft.world.World.setBlock(SourceFile:3318) ~[amu.class:?]
        at org.spongepowered.api.world.extent.MutableBlockVolume.setBlockType(MutableBlockVolume.java:96) ~[MutableBlockVolume.class:1.12.2-7.3.0]
        at dk.itu.real.ooe.services.MinecraftService.lambda$spawnBlocks$0(MinecraftService.java:55) ~[MinecraftService.class:?]
        at org.spongepowered.api.scheduler.Task$Builder.lambda$execute$0(Task.java:139) ~[Task$Builder.class:1.12.2-7.3.0]
        at org.spongepowered.common.scheduler.SchedulerBase.lambda$startTask$0(SchedulerBase.java:197) ~[SchedulerBase.class:1.12.2-7.3.0]
        at org.spongepowered.common.scheduler.SyncScheduler.executeTaskRunnable(SyncScheduler.java:74) ~[SyncScheduler.class:1.12.2-7.3.0]
        at org.spongepowered.common.scheduler.SchedulerBase.startTask(SchedulerBase.java:188) ~[SchedulerBase.class:1.12.2-7.3.0]
        at org.spongepowered.common.scheduler.SchedulerBase.processTask(SchedulerBase.java:174) ~[SchedulerBase.class:1.12.2-7.3.0]
        at java.util.concurrent.ConcurrentHashMap$ValuesView.forEach(Unknown Source) ~[?:1.8.0_291]
        at org.spongepowered.common.scheduler.SchedulerBase.runTick(SchedulerBase.java:112) ~[SchedulerBase.class:1.12.2-7.3.0]
        at org.spongepowered.common.scheduler.SyncScheduler.tick(SyncScheduler.java:47) ~[SyncScheduler.class:1.12.2-7.3.0]
        at org.spongepowered.common.scheduler.SpongeScheduler.tickSyncScheduler(SpongeScheduler.java:189) ~[SpongeScheduler.class:1.12.2-7.3.0]
        at net.minecraft.server.dedicated.DedicatedServer.handler$zpc000$vanilla$onTick(SourceFile:2277) ~[nz.class:?]
        at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(SourceFile:351) ~[nz.class:?]
        at net.minecraft.server.MinecraftServer.func_71217_p(SourceFile:560) ~[MinecraftServer.class:?]
        at net.minecraft.server.MinecraftServer.run(SourceFile:464) ~[MinecraftServer.class:?]
        at java.lang.Thread.run(Unknown Source) [?:1.8.0_291]
```
I'm not sure exactly what this is doing on the Python side. It might mean that the `spawnBlocks` command never returns which causes the Python program to freeze.

The key to understanding the error is this:
`expected in range (-30000000, 0, -30000000) to (29999999, 255, 29999999) but got (92, -1, 5)`

Basically, the y range is very narrow, and even with a limit on the number of blocks generated it is easy to attempt placing one at y-coordinate -1, resulting in an error.

The issue on confining the snakes #40 will address this, but I feel that there should be an additional case that prevents this crash even if the user does not want to confine the snakes. Basically, if the y coordinate is ever negative, then generation should immediately stop (set stop return value to True)

**Comment by richeyme** [comment]
So far it's been working pretty well. It has run for 13 generations without it crashing
